packages/task/src/browser/process/process-task-resolver.ts

```diff
@@ -36,19 +40,28 @@ export class ProcessTaskResolver implements TaskResolver {
         if (taskConfig.type !== 'process' && taskConfig.type !== 'shell') {
             throw new Error('Unsupported task configuration type.');
         }
+
+        const newRegistry = this.variableRegistry.getCopy();
+        newRegistry.registerVariable({
+            name: VAR_WORKSPACE_FOLDER,
```

Comment:
There is a defined semantic already for `workspaceFolder` variable. Why should it be changed for the task extension?

Reply:
In this particular case, I think it is because some tasks are bound to the workspace root that defines them, we can resolve the `workspaceFolder` variable to its value directly.

The default behavior to resolve this variable would be to ask the user which root to use (using quick open) if there is any ambiguity. But here it seems fair to assume that we know what value to resolve to.

Hence this local overriding of the variable, only the `newRegistry` in this function scope will be affected (it is kind of a snapshot where we override variables with more specific values for resolution).

Do you see any issues with that approach?

Reply:
I don't like changes to APIs, `VariableRegistry` was an implementation detail of `VariableResolverService`. There was a defined semantic for `workspaceFolder` value and the way how to contribute new variables.

> The default behavior to resolve this variable would be to ask the user which root to use (using quick open) if there is any ambiguity. But here it seems fair to assume that we know what value to resolve to.

It is not true. `workspaceFolder` is computed based on last opened file. The user won't be asked anything.

> Hence this local overriding of the variable, only the newRegistry in this function scope will be affected (it is kind of a snapshot where we override variables with more specific values for resolution).

If you run a task defined in one root in the context of a file from another root, why `workspaceRoot` should be resolved to a task root, not to a root of an opened file? The last is definition of `workspaceRoot`.

Reply:
>If you run a task defined in one root in the context of a file from another root, why workspaceRoot should be resolved to a task root, not to a root of an opened file? The last is definition of workspaceRoot.

Please correct me if I am wrong: Variable `workspaceFolder` resolves as the resource uri of the current widget, which is not necessarily the uri of the opened file. Therefore, without this change, a task could be run with the context that the user is unaware of.

Also, not all tasks are context-generic. For example, a task that builds a C++ project and involves a makefile cannot simply be started under any random folder. And I believe that happens to a lot of other scenarios.

Reply:
ok, what about the following approach, instead of exposing `VariableRegistry`:
- we add a section optional argument `context` which provides a URI of a current context:
```
interface VariableResolveOptions {
    context?: URI
}

resolve<T>(value: T, context: VariableResolveOptions = {}): Promise<T>
```
- we pass this context to `Variable.resolve`, i.e.
```ts
export interface Variable {
    resolve(context?: URI): MaybePromise<string | undefined>;
}
```
- we update `workspateRoot`, `workspaceFolder`, `workspaceFolderBasename` and `relativeFile` variables to respect context if it is given, if not they should work as before